### PR TITLE
functional-tester: add latency test (simulate slow network)

### PR DESCRIPTION
**Pull request by gyuho**
Fix https://github.com/coreos/etcd/issues/4666.

### CODE REVIEW

**Comment by gyuho** [comment]
Test results:

```
2016-03-03 19:44:02.241584 I | etcd-tester: [round#0 case#8] start failure slow down one member's network
2016-03-03 19:44:02.241623 I | etcd-tester: [round#0 case#8] start injecting failure...
2016-03-03 19:44:02.806642 I | etcd-tester: [round#0 case#8] injected failure
2016-03-03 19:44:02.806677 I | etcd-tester: [round#0 case#8] start recovering failure...
2016-03-03 19:44:05.680652 I | etcd-tester: [round#0 case#8] recovered failure
2016-03-03 19:44:05.680695 I | etcd-tester: [round#0 case#8] canceling the stressers...
2016-03-03 19:44:05.682762 I | etcd-tester: [round#0 case#8] canceled stressers
2016-03-03 19:44:05.682781 I | etcd-tester: [round#0 case#8] checking current revisions...
2016-03-03 19:44:06.750895 I | etcd-tester: [round#0 case#8] all members are consistent with current revisions
2016-03-03 19:44:06.750935 I | etcd-tester: [round#0 case#8] checking current storage hashes...
2016-03-03 19:44:06.750943 I | etcd-tester: [round#0 case#8] all members are consistent with storage hashes
2016-03-03 19:44:06.750949 I | etcd-tester: [round#0 case#8] restarting the stressers...
2016-03-03 19:44:06.750959 I | etcd-tester: [round#0 case#8] succeed!
2016-03-03 19:44:07.008208 I | etcd-tester: [round#0 case#9] start failure slow down leader's network
2016-03-03 19:44:07.008252 I | etcd-tester: [round#0 case#9] start injecting failure...
2016-03-03 19:44:07.650521 I | etcd-tester: [round#0 case#9] injected failure
2016-03-03 19:44:07.650565 I | etcd-tester: [round#0 case#9] start recovering failure...
2016-03-03 19:44:10.539886 I | etcd-tester: [round#0 case#9] recovered failure
2016-03-03 19:44:10.539919 I | etcd-tester: [round#0 case#9] canceling the stressers...
2016-03-03 19:44:10.542175 I | etcd-tester: [round#0 case#9] canceled stressers
2016-03-03 19:44:10.542204 I | etcd-tester: [round#0 case#9] checking current revisions...
2016-03-03 19:44:11.612462 I | etcd-tester: [round#0 case#9] all members are consistent with current revisions
2016-03-03 19:44:11.612507 I | etcd-tester: [round#0 case#9] checking current storage hashes...
2016-03-03 19:44:11.612528 I | etcd-tester: [round#0 case#9] all members are consistent with storage hashes
2016-03-03 19:44:11.612535 I | etcd-tester: [round#0 case#9] restarting the stressers...
2016-03-03 19:44:11.612549 I | etcd-tester: [round#0 case#9] succeed!
2016-03-03 19:44:11.831917 I | etcd-tester: [round#0 case#10] start failure isolate all members' network
2016-03-03 19:44:11.831950 I | etcd-tester: [round#0 case#10] start injecting failure...
2016-03-03 19:44:13.402634 I | etcd-tester: [round#0 case#10] injected failure
2016-03-03 19:44:13.402672 I | etcd-tester: [round#0 case#10] start recovering failure...
2016-03-03 19:44:18.100058 I | etcd-tester: [round#0 case#10] recovered failure
2016-03-03 19:44:18.100120 I | etcd-tester: [round#0 case#10] canceling the stressers...
2016-03-03 19:44:18.104559 I | etcd-tester: [round#0 case#10] canceled stressers
2016-03-03 19:44:18.104587 I | etcd-tester: [round#0 case#10] checking current revisions...
2016-03-03 19:44:19.172032 I | etcd-tester: [round#0 case#10] all members are consistent with current revisions
2016-03-03 19:44:19.172066 I | etcd-tester: [round#0 case#10] checking current storage hashes...
2016-03-03 19:44:19.172074 I | etcd-tester: [round#0 case#10] all members are consistent with storage hashes
2016-03-03 19:44:19.172080 I | etcd-tester: [round#0 case#10] restarting the stressers...
2016-03-03 19:44:19.172092 I | etcd-tester: [round#0 case#10] succeed!
```

**Comment by gyuho** [comment]
@xiang90 Please review. I just test-ran with 1-second latency:

```
2016-03-03 19:59:25.247615 I | etcd-tester: [round#0 case#8] start failure slow down one member's network
2016-03-03 19:59:25.247658 I | etcd-tester: [round#0 case#8] start injecting failure...
2016-03-03 19:59:26.238143 I | etcd-tester: [round#0 case#8] injected failure
2016-03-03 19:59:26.238186 I | etcd-tester: [round#0 case#8] start recovering failure...
2016-03-03 19:59:31.513282 I | etcd-tester: [round#0 case#8] recovered failure
2016-03-03 19:59:31.513339 I | etcd-tester: [round#0 case#8] canceling the stressers...
2016-03-03 19:59:31.516586 I | etcd-tester: [round#0 case#8] canceled stressers
2016-03-03 19:59:31.516601 I | etcd-tester: [round#0 case#8] checking current revisions...
2016-03-03 19:59:32.577718 I | etcd-tester: [round#0 case#8] all members are consistent with current revisions
2016-03-03 19:59:32.577764 I | etcd-tester: [round#0 case#8] checking current storage hashes...
2016-03-03 19:59:32.577774 I | etcd-tester: [round#0 case#8] all members are consistent with storage hashes
2016-03-03 19:59:32.577782 I | etcd-tester: [round#0 case#8] restarting the stressers...
2016-03-03 19:59:32.577792 I | etcd-tester: [round#0 case#8] succeed!
2016-03-03 19:59:32.810716 I | etcd-tester: [round#0 case#9] start failure slow down leader's network
2016-03-03 19:59:32.810745 I | etcd-tester: [round#0 case#9] start injecting failure...
2016-03-03 19:59:33.841371 I | etcd-tester: [round#0 case#9] injected failure
2016-03-03 19:59:33.841417 I | etcd-tester: [round#0 case#9] start recovering failure...
2016-03-03 19:59:35.524613 I | etcd-tester: [round#0 case#9] recovered failure
2016-03-03 19:59:35.524667 I | etcd-tester: [round#0 case#9] canceling the stressers...
2016-03-03 19:59:35.528627 I | etcd-tester: [round#0 case#9] canceled stressers
2016-03-03 19:59:35.528659 I | etcd-tester: [round#0 case#9] checking current revisions...
2016-03-03 19:59:36.592281 I | etcd-tester: [round#0 case#9] all members are consistent with current revisions
2016-03-03 19:59:36.592318 I | etcd-tester: [round#0 case#9] checking current storage hashes...
2016-03-03 19:59:36.592423 I | etcd-tester: [round#0 case#9] all members are consistent with storage hashes
2016-03-03 19:59:36.592431 I | etcd-tester: [round#0 case#9] restarting the stressers...
2016-03-03 19:59:36.592441 I | etcd-tester: [round#0 case#9] succeed!
2016-03-03 19:59:36.691494 I | etcd-tester: [round#0 case#10] start failure slow down all members' network
2016-03-03 19:59:36.691525 I | etcd-tester: [round#0 case#10] start injecting failure...
2016-03-03 19:59:39.811699 I | etcd-tester: [round#0 case#10] injected failure
2016-03-03 19:59:39.811735 I | etcd-tester: [round#0 case#10] start recovering failure...
2016-03-03 19:59:49.215268 I | etcd-tester: [round#0 case#10] recovered failure
2016-03-03 19:59:49.215343 I | etcd-tester: [round#0 case#10] canceling the stressers...
2016-03-03 19:59:49.224477 I | etcd-tester: [round#0 case#10] canceled stressers
2016-03-03 19:59:49.224520 I | etcd-tester: [round#0 case#10] checking current revisions...
2016-03-03 19:59:50.300953 I | etcd-tester: [round#0 case#10] all members are consistent with current revisions
2016-03-03 19:59:50.300995 I | etcd-tester: [round#0 case#10] checking current storage hashes...
2016-03-03 19:59:50.301003 I | etcd-tester: [round#0 case#10] all members are consistent with storage hashes
2016-03-03 19:59:50.301012 I | etcd-tester: [round#0 case#10] restarting the stressers...
2016-03-03 19:59:50.301023 I | etcd-tester: [round#0 case#10] succeed!
```

**Comment by gyuho** [comment]
Will revisit after v3 release.

**Comment by gyuho** [comment]
@xiang90 Just rebased. 

Haven't deployed testing cluster for awhile, and currently we are reaching 20k rounds.
A lot has changed since then. Do we want to test this as well in the next deployment?

Thanks!

File: tools/functional-tester/etcd-agent/agent.go [hunk 1]
```diff
@@ -172,6 +172,14 @@ func (a *Agent) recoverPort(port int) error {
 	return netutil.RecoverPort(port)
 }
 
+func (a *Agent) setLatency(ms, rv int) error {
+	return netutil.SetLatency(ms, rv)
+}
+
+func (a *Agent) removeLatency() error {
```
Comment (xiang90) [inline tools/functional-tester/etcd-agent/agent.go]: can we just set latency to 0? so we can remove this method?

File: tools/functional-tester/etcd-tester/failure.go [hunk 2]
```diff
@@ -295,3 +299,92 @@ func (f *failureIsolateAll) Recover(c *cluster, round int) error {
 	}
 	return c.WaitHealth()
 }
+
+type failureSlowNetwork struct {
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/failure.go]: SlowNetworkOneMember

File: tools/functional-tester/etcd-tester/failure.go [hunk 3]
```diff
@@ -295,3 +299,92 @@ func (f *failureIsolateAll) Recover(c *cluster, round int) error {
...
+	description
+}
+
+func newFailureSlowNetwork() *failureSlowNetwork {
+	return &failureSlowNetwork{
+		description: "slow down one member's network",
```
Comment (xiang90) [inline tools/functional-tester/etcd-tester/failure.go]: by how much? slow down one member's network by adding x latency

**Comment by xiang90** [comment]
LGTM.

File: tools/functional-tester/etcd-agent/agent.go (hunk 1 above)
Comment (gyuho) [inline tools/functional-tester/etcd-agent/agent.go]: ok will fix.

**Comment by gyuho** [comment]
@xiang90 Just fixed up. Will merge and deploy the testing cluster.

**Pull request merged**